ZOOKEEPER-3829: fix rolling restart when dynamic reconfig is disabled

symat opened the pull request:
In four different Jira tickets (ZOOKEEPER-3829, ZOOKEEPER-3830, ZOOKEEPER-3814, ZOOKEEPER-3842) we saw different errors when dynamic reconfig was disabled and we used rolling restart to change the quorum membership configuration. These rolling restart sequences was working fine on 3.4 but caused errors in 3.5 or 3.6.

In worst case the rolling restart leads to the scenario that we had an elected leader which was up but unable to commit any changes. This happens, when the quorum is extended with a new member in the following sequence:
* start server.1, server.2, server.3 (with config: 1,2,3)
* start server.4 (with config 1,2,3,4)
* stop server.1, then restart it with config 1,2,3,4
* stop server.2, then restart it with config 1,2,3,4
* stop server.3, then restart it with config 1,2,3,4
* at this point leader is server.4, but it can not commit any transaction

An other error we saw was when we changed a host name of an existing member (removing server.5 and add a new host as server.6). In this case we found in the logs of the new server (server.6) that it is still tried to connect to the old invalid server (server.5), although it was missing from it's config. The same problem remained even after making a full rolling-restart on all the nodes.

In this patch I try to fix these issues without breaking anything. The patch contains the following changes:
* We are making sure that neither the committed, nor the last seen config gets updated if dynamic reconfig is disabled.
* It is not possible now to start the leader without the ability of committing transaction, when dynamic reconfig is disabled (this is only needed to avoid a reconfig edge-case).
* I added a testcase simulating the enablement of dynamic reconfig using rolling restart
* I added a few more unit tests to cover rolling restart scenarios. (the tests are failing without the patch but succeeding after applying it).
* The enablement / disablement of reconfig is getting initialized now in the QuorumPeer and gets propagated to the other classes. This was needed for the rolling restart tests to be able to enable/disable reconfig only for the newly created servers without affecting the servers running already in the same JVM.

I also tested the changes with docker, using: https://github.com/symat/zookeeper-docker-test

target branches: 3.5, 3.6, master

eolivelli commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -1508,20 +1510,25 @@ private synchronized void startZkServer() {
+                allowedToCommit = false;
+            }
+        } else {
+            LOG.debug("Reconfig feature is disabled, skip designatedLeader calculation and reconfig processing.");
```
LOG.debug -> what about putting it as LOG.info ?

In your new test cases, are you exercising this branch or the "if QuorumPeerConfig.isReconfigEnabled()" branch ?

symat replied:
I agree with the LOG.info (this message will be printed only once when the leader starts to lead after an election. Yet, the message can help debugging production issues, when debug logs are disabled)

The tests are simulating the "backward compatibility" case when `QuorumPeerConfig.isReconfigEnabled()==false`. I don't think that changing the config with rolling-restart would be advised when dynamic-reconfig is enabled. I think when the reconfig is enabled, the new config can propagate during the leader election, so the rolling-restart will be unneccessary.

symat commented:
@shralex , @hanm please take a look if you can!

symat commented:
@lvfangmin, @fpj  - I think you also might have more background knowledge around the dynamic-reconfig. Can you please take a look on this PR?

shralex commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -614,7 +614,7 @@ void lead() throws IOException, InterruptedException {
             QuorumVerifier lastSeenQV = self.getLastSeenQuorumVerifier();
             QuorumVerifier curQV = self.getQuorumVerifier();
-            if (curQV.getVersion() == 0 && curQV.getVersion() == lastSeenQV.getVersion()) {
+            if (!QuorumPeerConfig.isReconfigEnabled() || (curQV.getVersion() == 0 && curQV.getVersion() == lastSeenQV.getVersion())) {
```
Could you explain the "!QuorumPeerConfig.isReconfigEnabled() || " here ? And not, e.g., "QuorumPeerConfig.isReconfigEnabled() &&"

The previous condition means this is done only when the config is an initial one. What's being done here is updating the config version to the current leader zxid. This is done only once, when the ensemble is initial.

symat replied:
Thanks for checking!

> What's being done here is updating the config version to the current leader zxid

I added the extra condition because here also the lastSeenQuorumVerifier is set to be equal to the currentQuorumVerifyer. The problem I found in ZOOKEEPER-3814 was that old server addresses were still propagated through the lastSeenQuorumVerifier to newly restarted nodes when dynamic reconfig was disabled. So my idea was to reset the lastSeenQuorumVerifier to the current config here. Maybe a better way would be to not update lastSeenQuorumVerifier when we process the NEWLEADER message (if dynamic reconfig is disabled)? 

> This is done only once, when the ensemble is initial.

Isn't the config version always 0 when dynamic reconfig is disabled?

symat replied:
I think I'll revert this part of the change in the next commit.

symat commented:
well, this is a very good question. Ideally I would say it should be enabled with a rolling restart. I think this is how people do all static config changes in production if the cluster must be up all time. We definitely should create a test for this, I doubt it exists. But Maybe we can do this in an other PR / ticket.

symat commented:
We had some chat on slack with @shralex and I will debug where the old server's configs are propagated (last seen config or last committed config) in the case of ZOOKEEPER-3814. I will try to avoid the update of both the committed / last seen config in case of disabled dynamic reconfig. I think that will solve ZOOKEEPER-3814.

I will also create (if there isn't one) a new automated test for the 'enablement of dynamic reconfig through rolling restart' to make sure we have that functionality.

symat commented:
In the latest commit I changed a few more things:
* I reverted the change questioned by @shralex [before](https://github.com/apache/zookeeper/pull/1356#discussion_r428720677)
* I followed all code paths where the committed, or the last seen config gets updated and made sure these never get changed when dynamic reconfig is disabled.
* I added a new testcase simulating the enablement of dynamic reconfig using rolling restart
* The enablement / disablement of reconfig is getting initialized now in the QuorumPeer and gets propagated to the other classes. This was needed for the rolling restart tests to be able to enable/disable reconfig only for the newly created servers without affecting the servers running already in the same JVM.
* I also rebased the PR on the current master branch

shralex commented:
I made a pass, it looks very good! One topic that I'm wondering about, is what happens when you enable / disable reconfig but 1) not all servers have the same value of the flag yet. For example, suppose only leader has reconfiguration enabled, it will accept a reconfig operation, and everyone else will say ACK but not process it. and 2) a leader boots and notices a .next file, indicating that a reconfig was in-progress. If the leader has reconfigEnabled=true, we're in case (1) - it will try to commit it via NEWLEADER/UPTODATE messages, and everyone may say ACK but not actually do anything about it. If the leader has reconfigEnabled=false, its possible that everyone else has the new config, but the leader won't commit it locally. It might be that these cases are OK, because if someone sets reconfigEnabled=false on a server, he expects the config to be local and that the config on different servers might diverge. Just wanted to bring this up.

shralex commented:
one of the new tests failed on Jenkins, FYI

symat commented on zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java:
```diff
@@ -943,6 +944,7 @@ public synchronized boolean tryToCommit(Proposal p, long zxid, SocketAddress fol
             self.processReconfig(newQV, designatedLeader, zk.getZxid(), true);
 
             if (designatedLeader != self.getId()) {
+                LOG.info("Committing a reconfiguration; this leader is not the designated leader anymore, setting allowedToCommit = false");
```
thanks, good idea

symat commented:
Thanks for going through the changes! I fixed your comments, also tried to make the failed test to be more robust. Based on the logs I think the problem was that I simply asserted a LOOKING state on a server which haven't recognised yet the loss of the leader. I changed the code to wait for the LOOKING state instead.

The other cases you mentioned about enabling / disabling rolling restart, I don't think they would be part of this 'backward-compatibility' issue. We can create a separate Jira for them. It would be great to detect these cases somehow and handle them (or at least put some error / warning to the logs). But honestly, I don't really think one should/would issue a rolling restart in parallel with an ongoing dynamic reconfig operation in production. I expect that enabling / disabling dynamic reconfig would be a very rare (ideally one-time) operation in a production cluster, issued very carefully. The second issue you describe might happen still sometime (having a .next file somehow beside the configs) when someone is re-enabling dynamic reconfig. I wonder if we shouldn't delete the .next files in the zkServer.sh / or in the java code during ZooKeeper startup when reconfigEnabled=false. This way "old" .next files would be less likely to left behind.

symat commented on zookeeper-server/src/test/java/org/apache/zookeeper/test/ReconfigExceptionTest.java:
```diff
@@ -90,6 +90,16 @@ public void tearDown() throws Exception {
     public void testReconfigDisabled() throws InterruptedException {
         QuorumPeerConfig.setReconfigEnabled(false);
+
+        // for thsi test we need to restart the quorum peers to get the config change,
```
thanks!

symat commented on zookeeper-server/src/test/java/org/apache/zookeeper/server/quorum/ReconfigRollingRestartCompatibilityTest.java:
```diff
@@ -209,6 +209,237 @@ public void testRollingRestartWithMembershipChange() throws Exception {
+                mt[i].start();
+                assertTrue("waiting for server " + i + " being up", ClientBase.waitForServerUp("127.0.0.1:" + clientPorts.get(i), CONNECTION_TIMEOUT));
+                verifyQuorumConfig(i, newServers, null);
+                verifyQuorumMembers(mt[i]);
```
It wouldn't hurt, I will add it (although the new leader shouldn't be selected from the first three servers). Still, a 'break' would make the code easier to follow for the reader.

symat commented:
the same test failed again, I have to check it. (it passes locally for me all the time, but it still must be flaky)

symat commented:
I removed the tests after I realized that this particularly restart sequence doesn't make sense. You first have to change the config of the existing nodes with rolling restart before starting the new nodes, otherwise (depending on the timing of the restarts and the number of new nodes) there can be separate partitions in the cluster as just happened with this test.

RajKiranS commented:
> should this be done with a rolling restart and if so, does the order of restart matter ? is there a test for something like this ?

Just noting down my observation when I enabled dynamic reconfig for the very first time. It was a bit scary! This is what I did for a 5 node ensemble:
1. Set `reconfigEnabled=true` in zoo.cfg on all the members.
2. Do a rolling-restart. First restart the followers and then the Leader at the end.

Restarted 2 of the 4 followers and it joined the cluster w/o any issues. And upon restarting the third follower, I could see leader emitting log messages saying that followers had different configuration. 💀 Then, I immediately restarted the leader and luckily the next leader chosen was a follower which already had `reconfigEnabled=true` set. And then restarted the remaining follower which safely joined the cluster. This was a scary experience.

Do not have the logs as its been a month or so I did this. But, I do remember that usual rolling-restart doesn't work as is. It needs to be ordered correctly. Like, update a follower, then next-likely-to-become-leader follower, leader, and then remaining followers. So, I think its best to test and document this procedure in detail.

nkalmar commented:
The change LGTM, I tested the patch applied to master with the quorum extension test. 
What I don't understand is what does backward compatibility have to do here? Both in the jira and here I don't see version upgrade mentioned. Same versions of ZK on every quorum member, right?

Sorry, this just doesn't add up, and I might be missing something :)

symat commented:
> What I don't understand is what does backward compatibility have to do here? 

By backward compatibility I meant that some rolling-restart sequences that was working well with 3.4 are not working with 3.5 and 3.6, even if `reconfigEnabled=false`. When the dynamic reconfig is disabled, then we would assume that rolling-restarts are behaving similarly as they did before 3.5. 

I think you are right, maybe this is a bad phrase here. I will try to make it more clear in the PR / Jira.

nkalmar commented:
It had 11 conflicting files, tried to merge them but I got a "Segmentation fault: 11" error... anyway, it didn't seem a trivial cherry-pick to 3.5, can you please create a PR against branch-3.5 @symat ? Merged to master and 3.6. Thanks!

symat commented:
Sure, thanks for the tests and the merge! I will create a new PR soon on 3.5 and also do some manual tests there.
